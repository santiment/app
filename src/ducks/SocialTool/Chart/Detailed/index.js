import React, { useState, useEffect, useMemo } from 'react'
import { graphql } from 'react-apollo'
import Chart from './Chart'
import { Metric } from '../../../dataHub/metrics'
import { PROJECT_METRICS_BY_SLUG_QUERY } from '../../../SANCharts/gql'
import { DetailedMetric } from './metrics'
import styles from './index.module.scss'

const GENERAL_CHARTS = [
  DetailedMetric.social_volume_telegram,
  DetailedMetric.social_volume_reddit,
  DetailedMetric.social_volume_twitter,
  DetailedMetric.social_volume_4chan,
]
const COMMUNITY_CHARTS = [DetailedMetric.community_messages_count_telegram]

const Colors = {}

GENERAL_CHARTS.forEach(({ key, color }) => (Colors[key] = color))
COMMUNITY_CHARTS.forEach(({ key, color }) => (Colors[key] = color))

const DefaultCharts = {
  general: GENERAL_CHARTS,
  community: [],
}

const DetailedBlock = ({
  linkedAssets,
  allDetectedAssets,
  MetricColor,
  priceAsset,
  settings,
  type,
  availableMetrics,
  children,
  ...props
}) => {
  const [MetricSettingMap, setMetricSettingMap] = useState(new Map())

  const charts = useMemo(
    () => {
      if (!availableMetrics) {
        return DefaultCharts[type]
      }

      return COMMUNITY_CHARTS.filter(({ key }) => availableMetrics.includes(key))
    },
    [availableMetrics],
    type,
  )

  const detectedAsset =
    type === 'community' ? allDetectedAssets.get(settings.slug) : linkedAssets.get(settings.slug)

  useEffect(() => {
    const newMetricSettingMap = new Map(MetricSettingMap)
    const metricSetting = {
      selector: detectedAsset ? 'slug' : 'text',
      slug: detectedAsset ? detectedAsset.slug : settings.slug,
    }

    if (charts.length > 0) {
      charts.forEach((metric) => newMetricSettingMap.set(metric, metricSetting))
      setMetricSettingMap(newMetricSettingMap)
    }
  }, [charts, settings.slug])

  if (charts.length === 0) {
    return null
  }

  const shouldShowChart = type !== 'community' || Boolean(detectedAsset)

  return !shouldShowChart || MetricSettingMap.size === 0 || !priceAsset ? null : (
    <>
      {charts.length > 0 && children}
      <div className={styles.charts}>
        {charts.map((chart) => (
          <div className={styles.chart} key={chart.key}>
            <span className={styles.label} style={{ '--color': chart.color }}>
              {chart.name}
            </span>
            <Chart
              {...props}
              settings={settings}
              colors={{ ...MetricColor, ...Colors }}
              metrics={[chart, { ...Metric.price_usd, reqMeta: { slug: priceAsset.slug } }]}
              MetricSettingMap={MetricSettingMap}
            />
          </div>
        ))}
      </div>
    </>
  )
}

export default graphql(PROJECT_METRICS_BY_SLUG_QUERY, {
  skip: ({ allDetectedAssets, settings, type }) => {
    const detectedAsset = allDetectedAssets.get(settings.slug)

    return !detectedAsset || type !== 'community' || settings.addedTopics.length > 0
  },
  options: ({ allDetectedAssets, settings }) => {
    const detectedAsset = allDetectedAssets.get(settings.slug)
    return { variables: { slug: detectedAsset.slug } }
  },
  props: ({ data: { project: { availableMetrics = [] } = {} } }) => {
    return { availableMetrics }
  },
})(DetailedBlock)
