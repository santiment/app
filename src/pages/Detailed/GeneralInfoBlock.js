import React from 'react'
import cx from 'classnames'
import { Icon } from '@santiment-network/ui'
import { formatCryptoCurrency, formatNumber } from './../../utils/formatting'
import HelpPopup from './../../components/HelpPopup/HelpPopup'
import styles from './GeneralInfoBlock.module.scss'

const GeneralInfoBlock = ({
  websiteLink,
  slackLink,
  twitterLink,
  githubLink,
  blogLink,
  whitepaperLink,
  marketcapUsd,
  rank,
  priceUsd,
  totalSupply,
  volumeUsd,
  ticker,
  roiUsd,
  isERC20
}) => (
  <div>
    <p>
      <SocialLink link={websiteLink} text='Website' />
      <SocialLink link={slackLink} text='Community' />
      <SocialLink link={twitterLink} text='Twitter' />
      <SocialLink link={blogLink} text='Blog' />
      <SocialLink link={githubLink} text='Github' />
      <SocialLink link={whitepaperLink} text='Whitepaper' />
    </p>
    <hr />
    <Row
      value={marketcapUsd}
      format={value => formatNumber(value, { currency: 'USD' })}
      title='Market Cap'
    />
    <Row
      value={priceUsd}
      format={value => formatNumber(value, { currency: 'USD' })}
      title='Price (USD)'
    />
    <Row
      value={volumeUsd}
      format={value => formatNumber(value, { currency: 'USD' })}
      title='Volume (USD)'
    />
    <Row
      value={totalSupply}
      format={value => formatCryptoCurrency(ticker, formatNumber(totalSupply))}
      title='Total supply'
    />
    <Row value={rank} format={value => value} title='Rank' />
    <Row
      value={roiUsd}
      format={value => parseFloat(value).toFixed(2)}
      title={
        <span>
          ROI since ICO
          <HelpPopup content='This ROI takes into account pre-sales, the token price during all sales and the amount of tokens distributed in each sale. Example: SAN had a pre-sale when around ~15m (12k ETH) tokens were distributed at a much lower price, and an ICO where the equivalent of 33k ETH were distributed. Both these sales are taken into account for this ROI, while most aggregators calculate ROI based only on the ICOâ€™s price.' />
        </span>
      }
    />
  </div>
)

const SocialLink = ({ link, text = '' }) => (
  <a className={styles.socialLink} href={link || ''}>
    {text || (
      <Icon type='link' fill={link ? 'var(--shark)' : 'var(--porcelain)'} />
    )}
  </a>
)

const DATA_IS_EMPTY = 'No data'

const Row = ({ title, value, format }) => (
  <div
    className={cx(
      'row-info',
      value === undefined || (!value && styles.disabled)
    )}
  >
    <div>{title}</div>
    <div>{!!value && value !== undefined ? format(value) : DATA_IS_EMPTY}</div>
  </div>
)

export default GeneralInfoBlock
