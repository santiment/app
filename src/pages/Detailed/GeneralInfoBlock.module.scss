.socialLink {
  font-size: 12px;
  padding: 0 10px;
}

.social-icons a {
  border-bottom: none;
  color: var(--jungle-green);
}

.disabled {
  color: var(--mystic);
}
