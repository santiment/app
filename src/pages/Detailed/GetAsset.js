import { graphql } from 'react-apollo'
import { projectBySlugGQL } from './DetailedGQL'
import { getTimeIntervalFromToday } from '../../utils/dates'

const GetAsset = ({ render, ...props }) => render({ ...props })

export default graphql(projectBySlugGQL, {
  name: 'Project',
  props: ({ Project }) => {
    const { projectBySlug = {} } = Project
    return {
      isLoading: Project.loading,
      isEmpty: !Project.hasOwnProperty('projectBySlug'),
      isError: Project.error,
      errorMessage: Project.error ? Project.error.message : '',
      project: projectBySlug
    }
  },
  options: ({ slug }) => {
    const { from: fromOverTime, to } = getTimeIntervalFromToday(-12 * 5, 'm')
    const { from } = getTimeIntervalFromToday(-1, 'm')
    return {
      variables: { slug, from, to, fromOverTime, interval: '7d' }
    }
  }
})(GetAsset)
