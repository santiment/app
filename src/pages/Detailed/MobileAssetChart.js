import React from 'react'
import moment from 'moment'
import {
  ResponsiveContainer,
  ComposedChart,
  Line,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
  ReferenceLine
} from 'recharts'
import { formatNumber } from './../../utils/formatting'
import styles from './MobileAssetChart.module.scss'

const MobileAssetChart = ({ data, slug: asset, icoPrice }) => {
  return (
    <div>
      <ResponsiveContainer width='100%' height={300}>
        <ComposedChart data={data}>
          <XAxis
            dataKey='datetime'
            tickLine={false}
            tickMargin={5}
            minTickGap={100}
            tickFormatter={timeStr => moment(timeStr).format('DD MMM YY')}
          />
          <YAxis
            yAxisId='axis-price'
            mirror
            axisLine={false}
            domain={['dataMin', 'auto']}
            tickFormatter={priceUsd =>
              formatNumber(priceUsd, { currency: 'USD' })
            }
          />
          <CartesianGrid vertical={false} stroke='#ebeef5' />
          <Tooltip
            labelFormatter={date => moment(date).format('dddd, MMM DD YYYY')}
            formatter={(value, name) => {
              if (name === `${asset}/USD`) {
                return formatNumber(value, { currency: 'USD' })
              }
              return value
            }}
          />
          <Line
            type='linear'
            yAxisId='axis-price'
            name={asset + '/USD'}
            dot={false}
            strokeWidth={1.5}
            dataKey='priceUsd'
            stroke={'var(--jungle-green)'}
          />
          {icoPrice && (
            <ReferenceLine
              yAxisId='axis-price'
              strokeDasharray='3 3'
              stroke='black'
              y={icoPrice}
            />
          )}
        </ComposedChart>
      </ResponsiveContainer>
      {icoPrice && (
        <div className={styles.icoPrice}>
          {`ICO Price ${formatNumber(icoPrice, { currency: 'USD' })}`}
          <div className={styles.icoPriceLegend} />
        </div>
      )}
    </div>
  )
}

export default MobileAssetChart
