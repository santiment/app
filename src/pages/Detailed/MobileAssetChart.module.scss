.icoPrice {
  display: flex;
}

.icoPriceLegend {
  width: 20px;
  height: 1px;
  margin-top: 10px;
  margin-left: 10px;
  border: 1px dashed black;
}

//Hide first YAxis tick on MobileAssetChart :with-love:
:global(.recharts-cartesian-axis-tick:first-child) {
  display: none;
}
