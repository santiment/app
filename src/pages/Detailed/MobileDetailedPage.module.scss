.wrapper {
  height: calc(100vh - 50px);
}

.main {
  padding: 0 16px;
}

.ticker {
  color: var(--waterloo);
}

.priceBlock {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.timeRangeBlock {
  margin-bottom: 1rem;
  margin-top: 1rem;
}

.priceUsd {
  font-size: 29px;
}

.changes24 {
  font-size: 24px;
}
