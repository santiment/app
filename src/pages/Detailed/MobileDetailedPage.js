import React, { useState } from 'react'
import cx from 'classnames'
import { compose } from 'recompose'
import { graphql } from 'react-apollo'
import { Label, Selector } from '@santiment-network/ui'
import { DailyActiveAddressesGQL } from './DetailedGQL'
import { TRANSACTION_VOLUME_QUERY } from '../../ducks/GetTimeSeries/queries/transaction_volume_query'
import { NEWS_QUERY } from '../../components/News/NewsGQL'
import { calcPercentageChange, capitalizeStr } from './../../utils/utils'
import {
  DAY,
  getTimeIntervalFromToday,
  getIntervalByTimeRange
} from '../../utils/dates'
import { formatNumber } from './../../utils/formatting'
import MobileHeader from './../../components/MobileHeader/MobileHeader'
import PercentChanges from './../../components/PercentChanges'
import NewsSmall from '../../components/News/NewsSmall'
import PageLoader from '../../components/Loader/PageLoader'
import MobileMetricCard from '../../components/MobileMetricCard/MobileMetricCard'
import GetAsset from './GetAsset'
import GetTimeSeries from '../../ducks/GetTimeSeries/GetTimeSeries'
import MobileAssetChart from './MobileAssetChart'
import ShowIf from '../../components/ShowIf'
import GetWatchlists from '../../ducks/Watchlists/GetWatchlists'
import WatchlistsPopup from '../../components/WatchlistPopup/WatchlistsPopup'
import { mergeTimeseriesByKey } from '../../utils/utils'
import { addRecentAssets } from '../../utils/recent'
import styles from './MobileDetailedPage.module.scss'

const MobileDetailedPage = props => {
  const slug = props.match.params.slug
  const [timeRange, setTimeRange] = useState('6m')
  const [extraMetric, setExtraMetric] = useState()

  addRecentAssets(slug)

  const toggleExtraMetric = toggledMetric => {
    if (extraMetric && toggledMetric.name === extraMetric.name) {
      setExtraMetric(undefined)
    } else {
      setExtraMetric(toggledMetric)
    }
  }

  const timeRangeBlock = (
    <div className={styles.timeRangeBlock}>
      <Selector
        options={['1w', '1m', '3m', '6m', 'all']}
        onSelectOption={setTimeRange}
        defaultSelected={timeRange}
      />
    </div>
  )

  let transactionVolumeInfo
  const { transactionVolume } = props
  if (transactionVolume && transactionVolume.length === 48) {
    const yesterdayTransactionVolume = transactionVolume
      .slice(0, 24)
      .reduce((acc, { transactionVolume }) => acc + transactionVolume, 0)
    const todayTransactionVolume = transactionVolume
      .slice(-24)
      .reduce((acc, { transactionVolume }) => acc + transactionVolume, 0)
    const TVDiff = calcPercentageChange(
      yesterdayTransactionVolume,
      todayTransactionVolume
    )
    transactionVolumeInfo = {
      name: 'Transaction Volume',
      value: todayTransactionVolume,
      period: '24h',
      changes: TVDiff
    }
  }

  let activeAddressesInfo
  const { dailyActiveAddresses } = props
  if (dailyActiveAddresses && dailyActiveAddresses.length === 2) {
    const [
      { activeAddresses: yesterdayActiveAddresses },
      { activeAddresses: todayActiveAddresses }
    ] = dailyActiveAddresses
    const DAADiff = calcPercentageChange(
      yesterdayActiveAddresses,
      todayActiveAddresses
    )
    activeAddressesInfo = {
      metric: 'dailyActiveAddresses',
      name: 'Daily Active Addresses',
      value: todayActiveAddresses,
      period: '24h',
      changes: DAADiff
    }
  }

  const { from, to } = getIntervalByTimeRange(timeRange)

  const timeseriesOptions = {
    slug,
    timeRange,
    interval: timeRange === '1w' ? '2h' : timeRange === '1m' ? '8h' : '1d'
  }

  const extraTimeserie = extraMetric
    ? {
      [extraMetric.name]: timeseriesOptions
    }
    : {}

  return (
    <div className={cx('page', styles.wrapper)}>
      <GetAsset
        slug={slug}
        render={({ isLoading, slug, project }) => {
          if (isLoading) {
            return (
              <>
                <MobileHeader
                  showBack
                  title={<Title slug={slug} />}
                  goBack={props.history.goBack}
                />
                <PageLoader />
              </>
            )
          }

          const {
            ticker,
            percentChange24h,
            percentChange7d,
            devActivity30,
            devActivity60,
            priceUsd,
            icoPrice
          } = project

          let devActivityInfo
          if (devActivity30 && devActivity60) {
            const DADiff = calcPercentageChange(
              devActivity60 * 2 - devActivity30,
              devActivity30
            )
            devActivityInfo = {
              metric: 'devActivity',
              name: 'Development Activity',
              value: devActivity30,
              period: '30d',
              changes: DADiff
            }
          }
          return (
            <>
              <MobileHeader
                showBack
                title={<Title slug={slug} ticker={ticker} />}
                goBack={props.history.goBack}
              />
              <GetWatchlists
                render={({ isLoggedIn }) => (
                  <WatchlistsPopup
                    projectId={project.id}
                    slug={project.slug}
                    isLoggedIn={isLoggedIn}
                  />
                )}
              />
              <div className={styles.main}>
                <PriceBlock
                  changes24h={percentChange24h}
                  changes7d={percentChange7d}
                  priceUsd={priceUsd}
                />
                {timeRangeBlock}
                <GetTimeSeries
                  historyPrice={timeseriesOptions}
                  {...extraTimeserie}
                  render={({
                    historyPrice = { items: [] },
                    ...otherTimeseries
                  }) => {
                    if (historyPrice.isLoading) {
                      return 'Loading...'
                    }

                    const timeseries = [historyPrice.items]
                    if (
                      extraMetric &&
                      otherTimeseries[extraMetric.name] &&
                      otherTimeseries[extraMetric.name].items
                    ) {
                      timeseries.push(otherTimeseries[extraMetric.name].items)
                    }
                    return (
                      <>
                        <MobileAssetChart
                          data={mergeTimeseriesByKey({
                            timeseries
                          })}
                          slug={slug}
                          icoPrice={icoPrice}
                          extraMetric={extraMetric}
                        />
                        <div className={styles.metrics}>
                          {activeAddressesInfo && (
                            <MobileMetricCard
                              {...activeAddressesInfo}
                              slug={slug}
                              activeMetric={extraMetric}
                              onClick={toggleExtraMetric}
                              from={from}
                              to={to}
                            />
                          )}
                          {devActivityInfo && (
                            <MobileMetricCard
                              {...devActivityInfo}
                              slug={slug}
                              activeMetric={extraMetric}
                              onClick={toggleExtraMetric}
                            />
                          )}
                          {transactionVolumeInfo && (
                            <MobileMetricCard
                              {...transactionVolumeInfo}
                              measure={ticker}
                            />
                          )}
                        </div>
                        <ShowIf news>
                          {props.news && props.news.length > 0 && (
                            <>
                              <h3 className={styles.news__heading}>News</h3>
                              <NewsSmall data={props.news} />
                            </>
                          )}
                        </ShowIf>
                      </>
                    )
                  }}
                />
              </div>
            </>
          )
        }}
      />
    </div>
  )
}

const Title = ({ slug, ticker }) => (
  <>
    {capitalizeStr(slug)}{' '}
    {ticker && <span className={styles.ticker}>({ticker.toUpperCase()})</span>}
  </>
)

const PriceBlock = ({ changes24h, changes7d, priceUsd }) => (
  <div>
    <div className={styles.priceUsd}>
      {priceUsd && formatNumber(priceUsd, { currency: 'USD' })}
    </div>
    <PercentChanges className={styles.changes} changes={changes24h} />
    <Label className={styles.label} accent='waterloo'>
      24h
    </Label>
    <PercentChanges className={styles.changes} changes={changes7d} />
    <Label className={styles.label} accent='waterloo'>
      7d
    </Label>
  </div>
)

const enhance = compose(
  graphql(NEWS_QUERY, {
    options: ({ match }) => {
      const tag = match.params.slug
      const { from, to } = getTimeIntervalFromToday(-3, DAY)
      return {
        variables: { from, to, tag, size: 6 }
      }
    },
    props: ({ data: { news = [] } }) => ({ news: news.reverse() })
  }),
  graphql(TRANSACTION_VOLUME_QUERY, {
    options: ({ match }) => {
      const to = new Date()
      const from = new Date()
      to.setUTCHours(to.getUTCHours() - 1, 0, 0, 0)
      from.setUTCHours(from.getUTCHours() - 48, 0, 0, 0)
      const { slug } = match.params
      return {
        variables: {
          slug,
          from: from.toISOString(),
          to: to.toISOString(),
          interval: '1h'
        }
      }
    },
    props: ({ data: { transactionVolume = [] } }) => ({ transactionVolume })
  }),
  graphql(DailyActiveAddressesGQL, {
    options: ({ match }) => {
      const { slug } = match.params
      const { from, to } = getTimeIntervalFromToday(-1, DAY)
      return { variables: { from, to, slug } }
    },
    props: ({ data: { dailyActiveAddresses = [] } }) => ({
      dailyActiveAddresses
    })
  })
)

export default enhance(MobileDetailedPage)
