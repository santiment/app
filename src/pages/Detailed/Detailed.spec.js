/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'
import toJson from 'enzyme-to-json'
import { Provider } from 'react-redux'
import Detailed from './Detailed'

jest.mock('../../components/Insight/comments/Comments.js', () => ({
  __esModule: true,
  default: () => <div />,
}))

jest.mock('../../components/Insight/PulseInsight/index.js', () => ({
  __esModule: true,
  default: () => <div />,
}))

describe('Project detail page container', () => {
  it('it should render correctly', () => {
    const store = configureStore([])({
      user: {
        data: {},
      },
      rootUi: {},
    })
    const match = {
      params: { ticker: 'AE' },
    }

    const pdp = shallow(
      <Provider store={store}>
        <Detailed
          project={{
            id: 12,
            name: 'Aragorn',
            ticker: 'AE',
            priceUsd: 10,
          }}
          match={match}
        />
      </Provider>,
    )

    expect(toJson(pdp)).toMatchSnapshot()
  })
})
