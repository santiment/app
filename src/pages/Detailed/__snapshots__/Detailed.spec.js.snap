// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project detail page container it should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <_default
    match={
      Object {
        "params": Object {
          "ticker": "AE",
        },
      }
    }
    project={
      Object {
        "id": 12,
        "name": "Aragorn",
        "priceUsd": 10,
        "ticker": "AE",
      }
    }
  />
</ContextProvider>
`;
